### PR TITLE
Adding sort=true keyword to readdir()

### PR DESCRIPTION
**Pull request by carstenbauer**
See #24626.

### CODE REVIEW

File: base/file.jl [hunk 1]
```diff
@@ -401,11 +401,11 @@ struct uv_dirent_t
 end
 
 """
-    readdir(dir::AbstractString=".") -> Vector{String}
+    readdir(dir::AbstractString="."; sort::Bool=true) -> Vector{String}
 
-Returns the files and directories in the directory `dir` (or the current working directory if not given).
+Returns the files and directories in the directory `dir` (or the current working directory if not given). If `sort=false` the resulting list won't necessarily be sorted.
```
Comment (StefanKarpinski) [inline base/file.jl]: Can you wrap this? We're not very good about consistently wrapping text but this is getting a bit long.

**Review by StefanKarpinski — COMMENTED** [review]
With a test to verify that entries are sorted, and an entry in `NEWS.md` this should be good to go. Great PR!

File: test/file.jl [hunk 2]
```diff
@@ -959,6 +959,20 @@ cd(dirwalk) do
...
+camefrom = pwd()
+dirread = mktempdir()
+cd(dirread)
+for k in 1:10
+    touch(randstring())
+end
+@test issorted(readdir()) # Check sorted output (#24626)
+rm(dirread, recursive=true)
+cd(camefrom)
+
```
Comment (crstnbr) [inline test/file.jl]: Wasn't entirely sure how to write a sensible test here. Is this acceptable?

Comment (KristofferC) [reply test/file.jl]: I would write this as

```
mktempdir() do path
    cd(path) do
        for k in 1:10
            touch(randstring())
        end
        @test issorted(readdir())
    end
end
```

File: base/file.jl [hunk 3]
```diff
@@ -401,11 +401,12 @@ struct uv_dirent_t
 end
 
 """
-    readdir(dir::AbstractString=".") -> Vector{String}
+    readdir(dir::AbstractString="."; sort::Bool=true) -> Vector{String}
 
 Returns the files and directories in the directory `dir` (or the current working directory if not given).
+If `sort=false` the resulting list won't necessarily be sorted.
```
Comment (stevengj) [inline base/file.jl]: comma before "the"

File: NEWS.md [hunk 4]
```diff
@@ -136,6 +136,9 @@ Language changes
     backslashes and the end of the literal while 2n+1 backslashes followed by a quote encodes n
     backslashes followed by a quote character ([#22926]).
 
+  * `readdir()` output is now guaranteed to be sorted on any OS. The keyword `sort` allows to
+    optionally turn off explicit sorting ([#24626]).
```
Comment (stevengj) [inline NEWS.md]: "allows to" is not grammatical in this context … you need to add a subject, like "allows one to…".  (Apparently this is a common mistranslation from *ermöglicht* in German?)

A more natural phrasing might be: "A new `sort` keyword argument can be used to disable explicit sorting."

File: base/file.jl (hunk 3 above)
Comment (stevengj) [reply base/file.jl]: Maybe some comment on why someone would pass `sort=false`: e.g. add (after a comma) "but performance may improve slightly.

Comment (stevengj) [reply base/file.jl]: (On my machine, calling `sort!` slows down `readdir` by about 5%, for either ~40 entries or ~1000 entries.)

File: NEWS.md (hunk 4 above)
Comment (crstnbr) [reply NEWS.md]: I had *erlaubt es* in mind. I guess impersonal, complicated formulations are natural for Germans :)

File: base/file.jl (hunk 3 above)
Comment (crstnbr) [reply base/file.jl]: I had also thought about that. Thanks for the review!

**Review by stevengj — APPROVED** [review]

File: NEWS.md [hunk 5]
```diff
@@ -136,6 +136,9 @@ Language changes
     backslashes and the end of the literal while 2n+1 backslashes followed by a quote encodes n
     backslashes followed by a quote character ([#22926]).
 
+  * `readdir()` output is now guaranteed to be sorted on any OS. A new `sort` keyword argument can
+    be used to disable explicit sorting ([#24626]).
```
Comment (tkelman) [inline NEWS.md]: This isn't a language change, would be more suited for the "Library improvements" section

**Review by StefanKarpinski — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
How much CI breakage are we expecting these days? These all seem unrelated but that's not as much green as I'd like to see before merging. In particular, no Windows tests passing is unfortunate since the new tests would have passed previously on other OSes.

**Comment by Sacha0** [comment]
The Travis i686 failure is another instance of the presently common libgit2 failure I mentioned last night on slack/infrastructure. Other instances:

https://travis-ci.org/JuliaLang/julia/jobs/305354871
https://travis-ci.org/JuliaLang/julia/jobs/305373961
https://travis-ci.org/JuliaLang/julia/jobs/305354871
https://travis-ci.org/JuliaLang/julia/jobs/305535401

@omus mentioned that https://github.com/JuliaLang/julia/pull/24338 might help sort that failure out? Best!

**Comment by tkelman** [comment]
What kind of code are you (or @bjarthur) writing that's sensitive to the order here? A case sensitive sorting usually doesn't make sense on Windows.

**Comment by StefanKarpinski** [comment]
> A case sensitive sorting usually doesn't make sense on Windows.

Why not? The point of this change is so that the default behavior is more platform independent, not to match any system-specific behavior. If you really want the native sort order, you can request it by  doing `readdir(sort=false)` which clearly warns the user that they may not get names in order (even if they happen to on their current system).

**Comment by omus** [comment]
> The Travis i686 failure is another instance of the presently common libgit2 failure I mentioned last night on slack/infrastructure

I'm looking into it currently.

**Comment by tkelman** [comment]
File system behavior isn't platform independent, and it doesn't seem worth making the default slower to pretend it is here. When working with the file system, seems more direct to return what the system call gives you, and add a call to `sort!` in user code where this actually makes a difference. Isn't composability and doing the fast thing by default the general julian API style?

**Comment by rfourquet** [comment]
> Isn't composability and doing the fast thing by default the general julian API style?

Cf. https://github.com/JuliaLang/julia/issues/939#issuecomment-53303719

**Comment by StefanKarpinski** [comment]
> and it doesn't seem worth making the default slower to pretend it is here

How many file names do you expect to be sorting? The cost of sorting an array of short strings is going to be dwarfed by the cost of the `readdir` operation.

> Isn't composability and doing the fast thing by default the general julian API style?

Doing the correct thing by default where it is reasonably fast is the Julian style, if that's not possible, then do the fast thing by default and make the correct thing possible, if slower.

**Comment by tkelman** [comment]
Sorting in a platform-inappropriate way is arguably less correct behavior for a list of file names than respecting the platform's iteration order by default.

**Comment by stevengj** [comment]
> How many file names do you expect to be sorting? The cost of sorting an array of short strings is going to be dwarfed by the cost of the readdir operation.

I did a couple of quick benchmarks, and it seemed to be a 5% slowdown; surprisingly, I got about the same overhead for both a small directory (40 entries) or a large directory (1000 entries).

I don't have strong feelings about this, but returning the same documented order on all platforms seems a bit friendlier, and makes it easier to reproduce results across platforms (since most programmers are probably unaware that the sort order varies).

**Comment by crstnbr** [comment]
> Maybe someone else should try it too?

Running a quick benchmark on my machine, I find 9.8% slowdown for 50 entries and 11.1% for 1000 entries.

**Comment by StefanKarpinski** [comment]
> Sorting in a platform-inappropriate way is arguably less correct behavior for a list of file names than respecting the platform's iteration order by default.

Correct by what criterion?

5-10% slowdown for consistent cross-platform behavior seems quite acceptable with an easy opt-out if someone really cares about the slowdown.

**Comment by tkelman** [comment]
Correct by the criterion of this being filesystem specific behavior where there's no expectation or reason for things to behave exactly the same everywhere? Python's `os.listdir` doesn't sort by default, and julia has generally followed the python design for most things in this area.

**Comment by tkelman** [comment]
If "consistent cross-platform behavior" is always desirable no matter what, then things like `joinpath` should use forward slashes everywhere. That would have been a defensible choice, if it were consistently made early on in the design of julia's filesystem APIs, but it wasn't.

In order for this not to be a 5-10% regression for other functionality (things like `walkdir`), most of the existing internal calls to `readdir` would have to be changed to use `sort=false` anywhere the order hasn't mattered.

**Comment by StefanKarpinski** [comment]
"Because Python does it that way" doesn't strike me as a great reason. Why is it more correct to list directory entries in the order that your OS happens to hand them to you? These functions do whatever we define them to do. If we define them to list directories in alphabetical order, then that's what they do. The `joinpath` argument is a non-sequitur – what path separator is used affects what the OS sees, not what the Julia program sees so it's the opposite consistency argument. The `walkdir` function should also sort everything by default and should have a corresponding `sort` keyword argument to opt out of sorting and just iterate in platform-dependent order.

**Comment by tkelman** [comment]
Node and ruby don't sort either, as far as I can tell. Should every function that calls `readdir` have a sort keyword argument then?

**Comment by StefanKarpinski** [comment]
> Should every function that calls `readdir` have a sort keyword argument then?

If it wants to give the caller the option not to sort. What's the alternative? The design of having the caller explicitly sort the output doesn't avoid that problem.

**Comment by omus** [comment]
My personal opinion is that we leave the sorting as is and document this behavior as part of the `readdir` docstring. Those who need the order to be consistent can use `sort!(readdir(...))`.

**Comment by StefanKarpinski** [comment]
@omus, reasons for that opinion? The pro case is that having consistent cross-platform behavior is desirable. The counterarguments seems to be:

1. Performance cost (5-10%),
2. Applications may want the native sort order – it's unclear (to me) why though.

What's clear, however, is that applications "get used to" whatever sort order they get on the platforms they run on and many applications are not run on multiple platforms, so they can get aways with making platform-specific assumptions if we allow them to. There doesn't seem to be any clear case why native order is better (aside from the performance cost of sorting). If we leave native order as the default, then we cannot change this in the future without annoying people and breaking code that hasn't been tested across platforms. If we sort by default, then that's one more thing that works the same everywhere by default when people write programs in Julia. I can't see how that's not a win.

**Comment by omus** [comment]
From my experience when using `readdir` I don't rely on any particular ordering and instead treat the result as an iterable set. The only case where ordering should matter is when writing test cases. During testing reproduciblity is more important than performance for me so I don't mind at all incurring the small performance penalty.

With that said I would be fine with this being merged. To me though this is mostly a documentation issue.

**Comment by rfourquet** [comment]
I'm mildly in favor of defaulting to the native order (i.e.`sort=false`), because I would tend to expect `readdir` to be as thin a wrapper of a system call as possible, although I don't have a case where this would matter. But it's not clear for me why a sorted order is relevant here, except e.g. when displaying to the user, in which case I find it better style to explicitly call `sort!` (or `sort=true`) to make the intent clear.

Also, is it likely that returning an iterator instead of a `Vector` be considered in the future? in which case, sorting would not be an option anymore. Alternatively, a new function could be introduced, like Python's `os.scandir` (cf.  https://www.python.org/dev/peps/pep-0471/ and [this issue](https://bugs.python.org/issue11406) which requests that `os.listdir` return an iterator).

**Comment by stevengj** [comment]
There are lots of cases in which the ordering matters for more than appearance.  Imagine a case where you do:
```jl
for testfile in readdir("tests")
    runtest(testfile)
end
```
and it works fine on the developers machine, but on the user's machine it breaks because the order changes.   You could say that the developer should not have depended on the order (intentionally or not), but cross-platform changes like this might be annoying to debug.

**Comment by StefanKarpinski** [comment]
Possible applications where order matters:

* testing
* hashing the contents of a file tree
* writing a file browser widget
* anything user-facing

The real issue here as @stevengj mentions is that we do get guaranteed sorted order on some systems but users can silently get a different and possibly broken experience. The fact that it would *only* be broken on Windows makes this situation worse since it's often the case that all of a library's developers work on UNIX systems but many users are on Windows.

@rfourquet: if we're going to default to `sort=false` then we might as well just have people do `sort!(readdir())` since it's shorter anyway. I do understand that since `readdir()` is the name of a libc function this would be expected to be a thin wrapper – bit it already isn't since it's a call to libuv, which already provides a level of platform independence. I would be fine with renaming this to `listdir` to avoid that expectation.

**Comment by tkelman** [comment]
For a browser or anything user facing, the platform order is much more natural. The function doesn't return things in a completely random order on Windows, it returns things in an order that's consistent with how nearly all other tools on the platform display files - sorted but case insensitive. It would be somewhat unnatural and surprising on Windows to group file names that start with an upper case letter first.

**Comment by rfourquet** [comment]
> if we're going to default to sort=false then we might as well just have people do sort!(readdir()) since it's shorter anyway

True, but functions like `walkdir` would still need to get this `sort` keyword.   

> I would be fine with renaming this to listdir to avoid that expectation.

I think I would much prefer the name `listdir` if it's sorted by default.

> hashing the contents of a file tree

I myself wrote such a script (with caching) to compare efficiently tree duplicates! And I can't imagine forgetting to sort the lists explicitly; but I'm using unix, so admitedly I would not expect to have a reliable order.

**Comment by StefanKarpinski** [comment]
> but I'm using unix, so admitedly I would not expect to have a reliable order.

But this is precisely the problem: on UNIX you *do* get a reliable order. So it's easy to not realize that you may not get the same order elsewhere.

> For a browser or anything user facing, the platform order is much more natural.

That's debatable. If someone really wants native ordering, they can just pass `sort=false`. A big part of the point of the runtime in a language is to abstract over the system details and present a consistent platform on which to run programs. In Julia code, there is a canonical ordering for strings – sorted lexicographically by code point (or more generally for invalid UTF-8, sorted by `memcmp`).

**Comment by rfourquet** [comment]
> But this is precisely the problem: on UNIX you do get a reliable order

Ah OK... but not reproducible, in that if a directory A is copied into B, then the list of files obtained in A won't match that of B, so for getting  reliable "hash signature" of a directory tree (so that A and B have the same signature), I need to sort the entries.

**Comment by tkelman** [comment]
If this were a serious problem that really warranted platform abstraction, then other languages would be sorting rather than following the platform conventions. This isn't just any arbitrary list of strings that get returned, it's a list of file names reflecting file system contents. You're not going to get case collisions on Windows where the Julia sorting conventions for strings are necessary or appropriate - the appropriate sorting function for a list of file names on Windows is one that ignores case.

> That's debatable

Does anyone else on this thread actually use windows? What's debatable about following platform conventions? UI toolkits that follow platform look and feel have proven far more popular and useful than ones that attempt to abstract everything into their own opinion about the single correct way to do things regardless of where they're run.

**Comment by StefanKarpinski** [comment]
There’s nothing preventing UI toolkits from following the native sorting. You’ve homed in on the one use case that can be argued either way here. That doesn’t address testing issues, hashing, reproducibility, etc.

The argument that other languages don’t abstract this platform behavior so we shouldn’t is pretty thin. Open source scripting languages like Python have traditionally not supported Windows well at all and made decisions like this *long* before it was a priority.

**Comment by tkelman** [comment]
Other languages have been around a very long time, and have had ample opportunity to make a change corresponding to this if they had ever deemed it worth doing. Python 2 to Python 3 made far more consequential changes than this one, but this issue didn't end up included as something they wanted to change during that, or any previous, transitions.

This isn't a matter of "supporting windows better" - it would be overriding the behavior that's most appropriate and conventional for that platform. I don't think that's a desirable default.

**Comment by crstnbr** [comment]
I guess I would argue that most users might not really care or properly know about platform specific order. On windows, the only case I can think of where the average user might really notice windows order is in the file explorer. But, I don't think that most users would really think of this as "windows order" but more (weird) "windows explorer order". In this sense, people won't expect windows order in Julia. They expect the output to be ordered or not - because the concept of windows and linux order probably does not exists in their heads.

However, this PR wasn't really my "idea".  I just wanted to contribute and looked for "good first issue" tags and created 2 PRs.

BTW: @tkelman I do use windows.

**Comment by tkelman** [comment]
It's not a weird ordering specific to explorer, it's a case insensitive order for a case insensitive file system, in every program and essentially every programming language that does this operation of listing files.

**Comment by PallHaraldsson** [comment]
The PR seems good to me (with sort by default), but it could be skipped on platforms that actually do give you sorted order(does that happen for all filesystems?).

**Comment by vtjnash** [comment]
One thing I would like to clarify is that the operating system doesn't have an order. Talking about "windows order" and "linux order" aren't logical, since they don't have a canonical ordering. Even "Window's explorer order" is ambiguous, since I don't know if you're talking about pre or post Windows  XP or what your group policy setting is (nor are the rules well documented: specifically, they say that it may change). You could talk about the filesystem listing order, but in many cases, that's effectively the same as unordered.

**Comment by stevengj** [comment]
Doing `let x = readdir(d); issorted(x) ? x : sort!(x); end` instead of `sort!(readdir(d))` speeds things up on my machine (MacOS) where the directory listing is already sorted — 3% slowdown instead of 7% slowdown vs. `readdir` for a directory with 1000 entries.

Wouldn't hurt to add this optimization, at least on non-Windows systems where `issorted(x)` will commonly return `true`.

**Comment by StefanKarpinski** [comment]
Bump. I still think this is a good idea and the overhead is negligible and you can opt out of it. Yes, the user can explicitly sort the output of `readdir` but (a) you have to know that you need to do that in order to get consistent results across platforms, and (b) that doesn't help with `walkdir`.

**Comment by JeffBezanson** [comment]
It's still not clear to me why readdir needs to give names in the same order on all platforms, when for example Dict iteration order isn't specified even on the same platform at different times. In other words, it's fine to have constructs with this kind of ordering behavior.

There also isn't a canonical "correct" sort order. We would be inserting our own opinion about how to sort strings, instead of just doing what other tools on the same platform do.

**Comment by StefanKarpinski** [comment]
Well, I want Dicts to preserve order too 😁

**Comment by crstnbr** [comment]
Given the discussion above, it's probably best to just add the `sort` keyword and keep `false` as the default, right? I just want to get some feedback before I finish this up.

**Comment by stevengj** [comment]
@crstnbr, if `sort=false` is the default, there's no point — as @StefanKarpinski commented above, you might as well type `sort!(readdir(...))` in that case.

**Comment by StefanKarpinski** [comment]
Reopen? https://mobile.twitter.com/bmarwell/status/1181821143835201536

**Pull request closed without merging**